[SPARK-21492][SQL][WIP] Memory leak in SortMergeJoin

taosaildrone opened the pull request:
## What changes were proposed in this pull request?

If SortMergeJoinScanner doesn't consume the iterator from
UnsafeExternalRowSorter entirely, the memory that
UnsafeExternalSorter acquired from TaskMemoryManager will not
be released. This leads to a memory leak, spills, and OOME. A
page will be held per partition of the unused iterator.
This patch will allow the SortMergeJoinScanner to explicitly close the iterators (for non-generated code)

## How was this patch tested?

Manual testing and profiling with scripts in SPARK-21492 comments.

taosaildrone committed:
[SPARK-21492][SQL][WIP] Memory leak in SortMergeJoin

If SortMergeJoinScanner doesn't consume the iterator from
UnsafeExternalRowSorter entirely, the memory that
UnsafeExternalSorter acquired from TaskMemoryManager will not
be released. This leads to a memory leak, spills, and OOME. A
page will be held per partition of the unused iterator.
This patch will allow the SortMergeJoinScanner to explicitly close the iterators (for non-generated code)

Manual testing and profiling with scripts in SPARK-21492 comments.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:
```diff
@@ -1140,9 +1167,26 @@ private class FullOuterIterator(
+
+trait CloseableScanner {
+  def closeIterator(iter: RowIterator): Unit = {
+    if (iter.isInstanceOf[RowIteratorFromScala]) {
```
I think you can use a `case` statement here to make it a little more concise, but no big deal.
Does this need to be a trait vs just a method that these various classes can call?

taosaildrone replied:
Case statement's a good idea. Looks cleaner.

Do you mean a class instead of a trait? The various classes are all top level, so I'm not sure where I'd put this method instead.

srowen replied:
I guess I figured this is just a static method in a companion object somewhere. If there really isn't a meaningfully cleaner place, this is OK.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:
```diff
@@ -1140,9 +1167,26 @@ private class FullOuterIterator(
+  def closeIterator(iter: RowIterator): Unit = {
+    if (iter.isInstanceOf[RowIteratorFromScala]) {
+      val rowIter = iter.asInstanceOf[RowIteratorFromScala]
+      val underlyingIter = rowIter.toScala
```
Does `.toScala` return the underlying thing or make a wrapper? Just checking, haven't investigated it, but want to make sure this closes what we think.

taosaildrone replied:
For our case of RowIteratorFromScala, this would return the underlying iterator

srowen commented on sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowIterator.java:
```diff
@@ -0,0 +1,26 @@
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow;
+import scala.collection.AbstractIterator;
+
+public abstract class UnsafeExternalRowIterator extends AbstractIterator<UnsafeRow> {
```
This should maybe extend Closeable?

srowen replied:
... and now you can just copy in the implementation that was previously in the anonymous subclass below? that might be cleaner overall. Hm, but maybe it won't work out because it has to call cleanupResources(). That part may have to stay in an anonymous subclass.

taosaildrone replied:
Done

srowen reviewed (changes requested):
I think this is a good idea in principle; I don't know this code well.

taosaildrone committed:
[SPARK-21492][SQL] Memory leak in SortMergeJoin

If SortMergeJoinScanner doesn't consume the iterator from
UnsafeExternalRowSorter entirely, the memory that
UnsafeExternalSorter acquired from TaskMemoryManager will not
be released. This leads to a memory leak, spills, and OOME. A
page will be held per partition of the unused iterator.
This patch will allow the SortMergeJoinScanner to explicitly close the iterators (for non-generated code)

Manual testing and profiling with scripts in SPARK-21492 comments.

taosaildrone commented:
Thanks for reviewing @srowen 

@kiszk @tejasapatil, would appreciate it if you have any input on how to approach the code-generation version of the fix.

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowIterator.java:
```diff
@@ -0,0 +1,74 @@
+import java.io.Closeable;
+import java.io.IOException;
+
+public abstract class UnsafeExternalRowIterator extends AbstractIterator<UnsafeRow> implements Closeable {
```
so we need an iterator API with a `close` method. Shall we reuse `RowIterator` and add `close` method to it?

taosaildrone replied:
I think this would require introducing a dependency on the spark-sql module from catalyst. 
And would require more refactoring of RowIterator and its subclasses.

hvanhovell commented:
@taosaildrone the sorted iterators consumed by SMJ should be cleaned up when the task finishes. Do you have any idea why this is not working?

cloud-fan commented:
IIUC the sorted iterator will be cleaned up eventually, this PR just clean it up earlier. I think this is not a memory leak fix, but an optimization.

taosaildrone commented:
There's a task completion listener, but 1) we could hit an OOME before the task completes or 2) impact performance by holding unnecessary memory and causing a bunch of unneeded spills. 
It's holding onto memory that is no longer needed, reducing available memory to the point that the application fails when it tries to allocate other memory necessary. I would classify that as a memory leak.

Something as simple as joining a DF with 1000 rows with another DF of 2 rows (with one overlapping row_id) will cause an OOME (both locally, and on a cluster) and cause the job to fail:

An example locally: ./bin/pyspark --master local[10] (tested on 2.4.0 and 3.0.0-master)
`

from pyspark.sql.functions import rand, col

spark.conf.set("spark.sql.join.preferSortMergeJoin", "true")
spark.conf.set("spark.sql.autoBroadcastJoinThreshold", -1)

r1 = spark.range(1, 1001).select(col("id").alias("timestamp1"))
r1 = r1.withColumn('value', rand())
r2 = spark.range(1000, 1001).select(col("id").alias("timestamp2"))
r2 = r2.withColumn('value2', rand())
joined = r1.join(r2, r1.timestamp1 == r2.timestamp2, "inner")
joined = joined.coalesce(1)
joined.explain()
joined.show()
`

hvanhovell commented:
@taosaildrone I have two problems with this approach:

1. This only works for a SMJ with Sorts as its direct input. That is the common case, however this won't work for any situation when there are operators in between the SMJ and the sort (for example when the sort is used between operators). 
2. I am not sure if it safe to assume that you can close an underlying child like this. You are toast as soon as something used in a downstream operator still points to a sort buffer you are releasing. This should generally not be an issue because it would probably has surfaced as a bug. However this is a pretty low bar to clear as it also depends on usage patterns. I would like to see a more principled approach here where you have both improved API (close on iterators???) and well defined semantics.

We should discuss this on the dev list.

taosaildrone commented:
@srowen @hvanhovell @cloud-fan 
Thanks for taking a look. I'm not tied to this approach, in fact hopefully someone with more knowledge in the Spark SQL internals can take this up. Here's the thread on the dev-list: 

http://apache-spark-developers-list.1001551.n3.nabble.com/Memory-leak-in-SortMergeJoin-td27152.html
